File: frontend/src/types/manifest.ts
export interface Source {
  id: string;
  name: string;
  regulatory_body: string;
  type: SourceType;
  format: SourceFormat;
  authority: AuthorityLevel;
  jurisdiction: Jurisdiction;
  url: string;
  access_method: AccessMethod;
  update_frequency?: string;
  last_known_update?: string;
  estimated_size?: string;
  scraping_notes?: string;
  confidence: number;
  needs_human_review: boolean;
  review_notes?: string;
  classification_tags: string[];
  relationships: Record<string, unknown>;
}

export interface RegulatoryBody {
  id: string;
  name: string;
  jurisdiction: Jurisdiction;
  authority_type: string;
  url: string;
  governs: string[];
}

export interface CoverageAssessment {
  total_sources: number;
  by_jurisdiction: Record<string, number>;
  by_type: Record<string, number>;
  completeness_score: number;
  known_gaps: KnownGap[];
}

export interface KnownGap {
  description: string;
  severity: "high" | "medium" | "low";
  mitigation: string;
}

export interface ManifestSummary {
  id: string;
  domain: string;
  status: ManifestStatus;
  created: string;
  sources_count: number;
  programs_count: number;
  coverage_score: number;
}

export interface Program {
  id: string;
  canonical_id: string;
  name: string;
  administering_entity: string;
  geo_scope: "national" | "state" | "county" | "city" | "tribal";
  jurisdiction?: string;
  benefits?: string;
  eligibility?: string;
  status: "active" | "paused" | "closed" | "verification_pending";
  last_verified?: string;
  evidence_snippet?: string;
  source_urls: string[];
  provenance_links: Record<string, unknown>;
  confidence: number;
  needs_human_review: boolean;
}

export interface ManifestDetail extends ManifestSummary {
  sources: Source[];
  programs: Program[];
  domain_map: {
    regulatory_bodies: RegulatoryBody[];
    jurisdiction_hierarchy: unknown;
  };
  coverage_assessment: CoverageAssessment | null;
}

export interface GenerateRequest {
  domain_description: string;
  llm_provider: string;
  k_depth?: number;
  geo_scope?: "national" | "state" | "municipal";
  target_segments?: string[];
}

export interface GenerateResponse {
  manifest_id: string;
  status: string;
  stream_url: string;
}

export interface ReviewRequest {
  reviewer: string;
  notes: string;
}

export interface AgentStepEvent {
  step: string;
  status: string;
  message?: string;
  bodies_found?: number;
  sources_found?: number;
  relationships_mapped?: number;
  completeness_score?: number;
  bodies_processed?: number;
  total_bodies?: number;
  manifest_id?: string;
  total_sources?: number;
  coverage_score?: number;
}

export interface AgentProgressEvent {
  sources_found: number;
  bodies_processed: number;
  total_bodies: number;
}

export type ManifestStatus = "generating" | "pending_review" | "approved" | "active" | "archived";
export type SourceType = "statute" | "regulation" | "guidance" | "standard" | "educational" | "guide";
export type SourceFormat = "html" | "pdf" | "legal_xml" | "api" | "structured_data";
export type AuthorityLevel = "binding" | "advisory" | "informational";
export type Jurisdiction = "federal" | "state" | "municipal";
export type AccessMethod = "scrape" | "download" | "api" | "manual";

File: backend/app/schemas/manifest.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.config import settings


class GenerateManifestRequest(BaseModel):
    manifest_name: str
    llm_provider: str = settings.llm_provider
    llm_model: str | None = None
    instruction_text: str | None = None
    k_depth: int = Field(default=2, ge=1, le=4)
    geo_scope: Literal["national", "state", "municipal"] = "state"
    target_segments: list[str] = []


class GenerateManifestResponse(BaseModel):
    manifest_id: str
    status: str
    stream_url: str


class SourceResponse(BaseModel):
    id: str
    name: str
    regulatory_body: str
    type: str
    format: str
    authority: str
    jurisdiction: str
    url: str
    access_method: str
    update_frequency: str | None = None
    last_known_update: str | None = None
    estimated_size: str | None = None
    scraping_notes: str | None = None
    confidence: float
    needs_human_review: bool
    review_notes: str | None = None
    classification_tags: list[str] = []
    relationships: dict = {}
    citation: str | None = None
    depth_hint: str | None = None


class SourceCreate(BaseModel):
    name: str
    regulatory_body: str
    type: str
    format: str
    authority: str
    jurisdiction: str
    url: str
    access_method: str
    update_frequency: str | None = None
    last_known_update: str | None = None
    estimated_size: str | None = None
    scraping_notes: str | None = None
    confidence: float = Field(ge=0.0, le=1.0, default=1.0)
    needs_human_review: bool = False
    classification_tags: list[str] = []
    citation: str | None = None
    depth_hint: str | None = None


class SourceUpdate(BaseModel):
    name: str | None = None
    url: str | None = None
    type: str | None = None
    format: str | None = None
    authority: str | None = None
    access_method: str | None = None
    update_frequency: str | None = None
    estimated_size: str | None = None
    scraping_notes: str | None = None
    confidence: float | None = Field(None, ge=0.0, le=1.0)
    needs_human_review: bool | None = None
    review_notes: str | None = None
    classification_tags: list[str] | None = None


class RegulatoryBodyResponse(BaseModel):
    id: str
    name: str
    jurisdiction: str
    authority_type: str
    url: str
    governs: list[str] = []


class CoverageAssessmentResponse(BaseModel):
    total_sources: int
    by_jurisdiction: dict = {}
    by_type: dict = {}
    completeness_score: float
    known_gaps: list[dict] = []


class DomainMapResponse(BaseModel):
    regulatory_bodies: list[RegulatoryBodyResponse] = []
    jurisdiction_hierarchy: dict | list = []


class ManifestSummary(BaseModel):
    id: str
    domain: str
    status: str
    created: datetime
    sources_count: int
    programs_count: int = 0
    coverage_score: float


class ProgramResponse(BaseModel):
    id: str
    canonical_id: str
    name: str
    administering_entity: str
    geo_scope: str
    jurisdiction: str | None = None
    benefits: str | None = None
    eligibility: str | None = None
    status: str
    last_verified: datetime | None = None
    evidence_snippet: str | None = None
    source_urls: list[str] = []
    provenance_links: dict = {}
    confidence: float
    needs_human_review: bool


class GoldenProgramResponse(BaseModel):
    id: int
    domain: str | None = None
    golden_run_id: str | None = None
    merge_key: str
    canonical_id: str
    name: str
    administering_entity: str
    geo_scope: str
    jurisdiction: str | None = None
    benefits: str | None = None
    eligibility: str | None = None
    status: str
    last_verified: datetime | None = None
    evidence_snippet: str | None = None
    source_urls: list[str] = []
    provenance_links: dict = {}
    confidence: float
    needs_human_review: bool
    source_manifest_ids: list[str] = []
    found_by_count: int
    ensemble_confidence: float
    merged_at: datetime | None = None


class ManifestDetail(BaseModel):
    id: str
    domain: str
    status: str
    created: datetime
    sources_count: int
    programs_count: int = 0
    coverage_score: float
    sources: list[SourceResponse] = []
    programs: list[ProgramResponse] = []
    domain_map: DomainMapResponse = DomainMapResponse()
    coverage_assessment: CoverageAssessmentResponse | None = None
    checkpoint_data: dict | None = None
    run_params: dict | None = None


class ManifestListResponse(BaseModel):
    manifests: list[ManifestSummary]


class GoldenProgramListResponse(BaseModel):
    programs: list[GoldenProgramResponse]
    total: int


class GoldenProgramStatsResponse(BaseModel):
    total: int
    by_geo_scope: dict[str, int] = {}
    by_found_by_count: dict[str, int] = {}
    average_ensemble_confidence: float = 0.0


class LogicalRunResponse(BaseModel):
    run_id: str
    manifest_id: str
    domain: str
    status: str
    created_at: datetime
    promoted_to_golden_run_id: str | None = None
    notes: str | None = None


class LogicalRunListResponse(BaseModel):
    runs: list[LogicalRunResponse]


class GoldenRunSummaryResponse(BaseModel):
    golden_run_id: str
    domain: str
    version: int
    source_run_ids: list[str] = []
    accepted_at: datetime
    accepted_by: str
    notes: str | None = None
    strategy: str
    item_count: int
    is_current: bool = False


class GoldenRunListResponse(BaseModel):
    runs: list[GoldenRunSummaryResponse]


class GoldenRunDetailResponse(GoldenRunSummaryResponse):
    programs: list[GoldenProgramResponse] = []


class PromoteGoldenRunRequest(BaseModel):
    domain: str
    source_run_ids: list[str] = Field(min_length=1)
    accepted_by: str = "system"
    notes: str = ""


class PromoteGoldenRunResponse(BaseModel):
    golden_run_id: str
    domain: str
    version: int
    total_input: int
    unique_output: int
    duplicates_removed: int
    new_added: int
    updated: int
    by_geo_scope: dict[str, int] = {}


class MergeManifestsRequest(BaseModel):
    manifest_ids: list[str] = Field(min_length=1)


class MergeManifestsResponse(BaseModel):
    total_input: int
    unique_output: int
    duplicates_removed: int
    new_added: int
    updated: int
    by_geo_scope: dict[str, int] = {}


class ReviewRequest(BaseModel):
    reviewer: str
    notes: str = ""


class ReviewResponse(BaseModel):
    manifest_id: str
    status: str
    approved_at: datetime | None = None
    rejection_notes: str | None = None
